### DISCUSSION TITLE
Target Markdown Questions

### DISCUSSION THREAD

**Discussion by jennysjaarda** [ropensci/targets#620] [Help]
Hello, 

I just have a few questions about using Target Markdown. 

1. To clarify, everything that happens in a a typical `_targets.R` file, can instead be found in a Targets Markdown report - is that right? I already have my pipeline setup in a `_targets.R` script. So if I wanted to make my whole project more interactive, I could move all targets to a (or multiple) Target Markdown script?  
2. How does this integrate with HPC clusters? Could some of my targets be in `_targets.R` file and others in the markdown file(s)? 
3. I am using `workflowr` (which I see you mention) so I have multiple Rmarkdown reports, does location of these Rmd files matter so they can communicate and find the `_targets` folder? 
4. Lastly, I am constantly switching between my personal computer and a remote HPC cluster. I tend to manage and edit my Rmd reports on my personal computer before eventually pushing them to the server via git. Do you have any workflow suggestions for managing this type of project? I do find it a pain sometimes to be going back and forth between the two computers. My project is way too big to track via git, but the summary files and figures would be quite small. Is there a way to track certain targets via git (i.e. the ones we want to explore visually with our Rmd reports)? This could be quite useful for these type of workflows. 

Thanks in advance for your suggestions and advice!
Jenny

**Comment by wlandau** [comment]
> To clarify, everything that happens in a a typical _targets.R file, can instead be found in a Targets Markdown report - is that right?

Target Markdown is a fancy way of writing `_targets.R` (or other manually set target script path) and supporting scripts. So yes, it is a layer on top of `_targets.R` and everything in a Target Markdown code chunk is forwarded to on of the helper scripts that `_targets.R` depends on.

> So if I wanted to make my whole project more interactive, I could move all targets to a (or multiple) Target Markdown script?

Yes, moving to Target Markdown gives you interactive mode while still preserving the usual non-interactive behavior by forwarding chunks to scripts in `_targets_r` in non-interactive mode.

> How does this integrate with HPC clusters? 

In interactive mode, native HPC is not available. (If you are running interactive mode, the assumption is that the targets will run fast, e.g. you have scaled them down for testing/prototyping.) In non-interactive mode, Target Markdown only writes `_targets.R` etc., so this is before you invoke HPC. So you could have Target Markdown code chunks that define the pipeline as usual, and then you could invoke `tar_make_clustermq()` either in a normal R code chunk later on or outside the report. Target Markdown is independent from HPC but compatible. Example: https://github.com/wlandau/rmedicine2021-pipeline/blob/05833d39dd0a5dc4d53c006e5acef6b74c9da50f/index.Rmd#L239 (rendered document: https://wlandau.github.io/rmedicine2021-pipeline/).

>  Could some of my targets be in _targets.R file and others in the markdown file(s)?

Non-interactive Target Markdown completely replaces the existing `_targets.R` (and there are protections against accidentally overwriting it). If you use Target Markdown for a project, `_targets.R` is read-only. So all your targets would then go in code chunks split over one or more reports.

> I am using workflowr (which I see you mention) so I have multiple Rmarkdown reports, does location of these Rmd files matter so they can communicate and find the _targets folder?

If all the reports are part of the same pipeline, they should be in the same working directory, which allows them to talk to the same `_targets.R` file.

> Lastly, I am constantly switching between my personal computer and a remote HPC cluster. I tend to manage and edit my Rmd reports on my personal computer before eventually pushing them to the server via git. Do you have any workflow suggestions for managing this type of project? I do find it a pain sometimes to be going back and forth between the two computers. My project is way too big to track via git, but the summary files and figures would be quite small. Is there a way to track certain targets via git (i.e. the ones we want to explore visually with our Rmd reports)? This could be quite useful for these type of workflows.

You can download/commit `_targets/meta/meta` and the specific targets you need in  `_targets/objects/`. Then, `tar_read()` will work locally for exploratory purposes. Or, if there is a way to locally mount the remote drive on the server where your project lives, `tar_read(store = "...")` might be able to download things.

**Reply by altairwei** [reply]
> > I am using workflowr (which I see you mention) so I have multiple Rmarkdown reports, does location of these Rmd files matter so they can communicate and find the _targets folder?
> 
> If all the reports are part of the same pipeline, they should be in the same working directory, which allows them to talk to the same `_targets.R` file.

It's impossible to manage the dependencies between RMarkdown documents through Target Markdown, right? Because the pipeline is defined by the Targets Markdown, however, the execution order of the RMarkdown documents is not controlled by the Targets Markdown. As I understand it, Targets Markdown only works in the case of a single document, or is there something wrong with my understanding?

**Reply by wlandau** [reply]
Target Markdown does work for more than one R Markdown document. Target Markdown operates on the level of an individual code chunk, and it acts on those code chunks independently, so you can distribute those chunks over however many documents you want. Then, you can run those documents in sequence to write the `_targets_r/` scripts for the pipeline (which can happen in any order) and then run `tar_make()` at the end (either in the last document or on its own).

Managing multiple R Markdown documents across multiple projects  got easier since I last commented on this thread, and I recommend having a look at https://books.ropensci.org/targets/projects.html. `tar_config_set()` modifies the YAML config files in the `TAR_CONFIG` environment variable, and the `TAR_PROJECT` environment variable lets you switch around multiple projects that may be defined in a single YAML. The top of each R Markdown document can select the correct config/project for the situation. The `root.dir` chunk option sets the working directory for the rest of the report, so that's useful here too.

**Reply by altairwei** [reply]
> Target Markdown does work for more than one R Markdown document. Target Markdown operates on the level of an individual code chunk, and it acts on those code chunks independently, so you can distribute those chunks over however many documents you want. Then, you can run those documents in sequence to write the `_targets_r/` scripts for the pipeline (which can happen in any order) and then run `tar_make()` at the end (either in the last document or on its own).

After you explained the working model of Target Markdown, I understood the mistake I made. In interactive mode, the Target Markdown chunks behave similarly to normal R chunks in that both are executed immediately in the current runtime environment. However, in non-interactive mode, Target Markdown chunks are actually just defining the pipeline, and the user codes in Target Markdown chunks do not run immediately, but rather wait until `tar_make()` is called. The mix of interactive and non-interactive modes explained in the tutorial led me to believe that both modes run as soon as RMarkdown is rendered, but in fact two modes run at completely different times. So, normal R chunks should not be mixed with Target Markdown chunks (one of the mistakes I made), and the former must wait for `tar_make()` to be called.

**Comment by jmbuhr** [comment]
It might also be possible to run rstudio server (https://www.rstudio.com/products/rstudio/download-server/) on the HPC cluster, which would allow you to work where your data lives from the comfort of your laptop and a browser, or the remote extension of VS Code would achieve something similar.
